# LogisticRegression Algorithm for handle a iris dataset
Six Sesemter Final_Lab_Exam Task on Machine Leaning Algorithm 

In [35]:
import pandas as pd
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
iris = load_iris()
iris_bunch = load_iris(as_frame=True)

In [16]:
iris.feature_names

['sepal length (cm)',
 'sepal width (cm)',
 'petal length (cm)',
 'petal width (cm)']

In [17]:
import matplotlib.pyplot as plt
%matplotlib inline

In [18]:
from sklearn.model_selection import train_test_split

In [20]:
# Convert to DataFrame
iris = iris_bunch.frame

# Check the columns in the DataFrame
print(iris.columns)

Index(['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)',
       'petal width (cm)', 'target'],
      dtype='object')


In [21]:
# Drop the 'target' column if it exists and store the result in X
# Note: Ensure 'flower_names' column is not being referenced if it doesn't exist
X = iris.drop(columns=['target'])  # Remove 'flower_names' if it does not exist

# Extract the 'target' column and store it in y
y = iris['target']

# Split the dataset into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


In [22]:
X_train

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
22                 4.6               3.6                1.0               0.2
15                 5.7               4.4                1.5               0.4
65                 6.7               3.1                4.4               1.4
11                 4.8               3.4                1.6               0.2
42                 4.4               3.2                1.3               0.2
..                 ...               ...                ...               ...
71                 6.1               2.8                4.0               1.3
106                4.9               2.5                4.5               1.7
14                 5.8               4.0                1.2               0.2
92                 5.8               2.6                4.0               1.2
102                7.1               3.0                5.9               2.1

[120 rows x 4 columns]

In [23]:
from sklearn.linear_model import LogisticRegression
model = LogisticRegression()

In [24]:
model.fit(X_train, y_train)

LogisticRegression()

In [25]:
X_test

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
73                 6.1               2.8                4.7               1.2
18                 5.7               3.8                1.7               0.3
118                7.7               2.6                6.9               2.3
78                 6.0               2.9                4.5               1.5
76                 6.8               2.8                4.8               1.4
31                 5.4               3.4                1.5               0.4
64                 5.6               2.9                3.6               1.3
141                6.9               3.1                5.1               2.3
68                 6.2               2.2                4.5               1.5
82                 5.8               2.7                3.9               1.2
110                6.5               3.2                5.1               2.0
12                 4.8               3.0                1.4               0.1
36                 5.5               3.5                1.3               0.2
9                  4.9               3.1                1.5               0.1
19                 5.1               3.8                1.5               0.3
56                 6.3               3.3                4.7               1.6
104                6.5               3.0                5.8               2.2
69                 5.6               2.5                3.9               1.1
55                 5.7               2.8                4.5               1.3
132                6.4               2.8                5.6               2.2
29                 4.7               3.2                1.6               0.2
127                6.1               3.0                4.9               1.8
26                 5.0               3.4                1.6               0.4
128                6.4               2.8                5.6               2.1
131                7.9               3.8                6.4               2.0
145                6.7               3.0                5.2               2.3
108                6.7               2.5                5.8               1.8
143                6.8               3.2                5.9               2.3
45                 4.8               3.0                1.4               0.3
30                 4.8               3.1                1.6               0.2

In [26]:
y_predicted = model.predict(X_test)

In [27]:
model.predict(X_test)

array([1, 0, 2, 1, 1, 0, 1, 2, 1, 1, 2, 0, 0, 0, 0, 1, 2, 1, 1, 2, 0, 2,
       0, 2, 2, 2, 2, 2, 0, 0])

In [28]:
model.predict_proba(X_test)

array([[3.78468964e-03, 8.27220592e-01, 1.68994718e-01],
       [9.46738570e-01, 5.32612296e-02, 2.00200608e-07],
       [8.73019002e-09, 1.55846967e-03, 9.98441522e-01],
       [6.43338014e-03, 7.92106296e-01, 2.01460324e-01],
       [1.44094299e-03, 7.74350518e-01, 2.24208539e-01],
       [9.55783509e-01, 4.42163143e-02, 1.77062516e-07],
       [7.76129022e-02, 9.08072540e-01, 1.43145575e-02],
       [1.61393845e-04, 1.55689752e-01, 8.44148855e-01],
       [2.20749609e-03, 7.62773187e-01, 2.35019317e-01],
       [2.83156446e-02, 9.45786638e-01, 2.58977173e-02],
       [4.39625408e-04, 2.43308395e-01, 7.56251980e-01],
       [9.68317751e-01, 3.16821713e-02, 7.81652986e-08],
       [9.72944229e-01, 2.70557379e-02, 3.33763996e-08],
       [9.62107119e-01, 3.78927703e-02, 1.11058752e-07],
       [9.79275316e-01, 2.07246192e-02, 6.48256053e-08],
       [4.54171482e-03, 7.12618594e-01, 2.82839692e-01],
       [7.22701932e-06, 2.42151312e-02, 9.75777642e-01],
       [2.73295948e-02, 9.47681

In [29]:
model.score(X_test,y_test)

1.0

In [30]:
model.score(X_train,y_train)

0.975

In [31]:
y_predicted

array([1, 0, 2, 1, 1, 0, 1, 2, 1, 1, 2, 0, 0, 0, 0, 1, 2, 1, 1, 2, 0, 2,
       0, 2, 2, 2, 2, 2, 0, 0])

In [32]:
X_test

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
73                 6.1               2.8                4.7               1.2
18                 5.7               3.8                1.7               0.3
118                7.7               2.6                6.9               2.3
78                 6.0               2.9                4.5               1.5
76                 6.8               2.8                4.8               1.4
31                 5.4               3.4                1.5               0.4
64                 5.6               2.9                3.6               1.3
141                6.9               3.1                5.1               2.3
68                 6.2               2.2                4.5               1.5
82                 5.8               2.7                3.9               1.2
110                6.5               3.2                5.1               2.0
12                 4.8               3.0                1.4               0.1
36                 5.5               3.5                1.3               0.2
9                  4.9               3.1                1.5               0.1
19                 5.1               3.8                1.5               0.3
56                 6.3               3.3                4.7               1.6
104                6.5               3.0                5.8               2.2
69                 5.6               2.5                3.9               1.1
55                 5.7               2.8                4.5               1.3
132                6.4               2.8                5.6               2.2
29                 4.7               3.2                1.6               0.2
127                6.1               3.0                4.9               1.8
26                 5.0               3.4                1.6               0.4
128                6.4               2.8                5.6               2.1
131                7.9               3.8                6.4               2.0
145                6.7               3.0                5.2               2.3
108                6.7               2.5                5.8               1.8
143                6.8               3.2                5.9               2.3
45                 4.8               3.0                1.4               0.3
30                 4.8               3.1                1.6               0.2

**model.coef_ indicates value of m in y=m*x + b equation**

In [33]:
model.coef_

array([[-0.39705946,  0.96066053, -2.37396368, -1.00330112],
       [ 0.51283652, -0.25352109, -0.21527075, -0.76918663],
       [-0.11577705, -0.70713943,  2.58923443,  1.77248775]])

**model.intercept_ indicates value of b in y=m*x + b equation**

In [34]:
model.intercept_

array([  9.03180018,   1.84145661, -10.87325679])

**Lets defined sigmoid function now and do the math with hand**

In [15]:
import math
def sigmoid(x):
  return 1 / (1 + math.exp(-x))

In [16]:
def prediction_function(age):
    z = 0.042 * age - 1.53 # 0.04150133 ~ 0.042 and -1.52726963 ~ -1.53
    y = sigmoid(z)
    return y

In [17]:
age = 35
prediction_function(age)

0.4850044983805899

**0.485 is less than 0.5 which means person with 35 age will *not* buy insurance**

In [18]:
age = 43
prediction_function(age)

0.568565299077705

**0.568 is more than 0.5 which means person with 43 will buy the insurance**